### PR TITLE
Plantmaster Mk4

### LINKED ISSUE #2273: Plantmaster Mk3 UI spam as borg
[BUG]

**Describe the bug**
On using the Plantmaster in botany as borg, everytime someone else uses the machine I get spammed with the UI popping up. Very very frustrating.
**To Reproduce**
1. Be borg.
2. Add fruit to plantmaster via satchel
3. Try to get on with growing things while a human uses the plantmaster
4. Get spammed to fuck with UI popups.

**Expected behavior**
To not be spammed.

### LINKED ISSUE #4974: PlantMaster Mk3 stops working, freezes game, then reconnects.
Submitter: De Luuk/Mr. Green
Describe the bug
When extracting seeds in a PlantMaster Mk3, the game freezes for a bit, then seems to reconnect. Clicking the PlantMaster again after that makes the same thing happen again. Other people clicking the PlantMaster once it's bugged experience the same issue. Since the reconnect closes the PlantMaster's screen any seeds or extractables are inaccessible.

To Reproduce
I have done some testing to find a way to consistently reproduce the bug, but was unable to find one. For example, in one test i set up three PlantMasters in the same way, but was only able to trigger the bug once. The steps below are the steps i used in that test.

1. Buy some produce at cargo. 
2. Load 250 extractables into a PlantMaster.
3. Extract seeds. Start at the bottom of the list. This should allow you to keep clicking without having to move your mouse much, if at all. Click at about the same rate the extractables disappear from the list.
4. Bug possibly happens.

Additional context:
Gerhazo asked me to mention there were no runtime errors.

When Gerhazo removed all the contents from a bugged PlantMaster, it was no longer bugged.

Whether or not seeds are retained in the machine does not seem relevant, nor does the amount of seeds in the machine or the type of extractable, based on my tests and previous occurences.

Reported on: Goonstation 1 Classic: Heisenbee 2021-05-31 13:34:48

### LINKED ISSUE #6429: Plantmaster Mk3 visually duplicate seeds on drag drop
### Labels

[BUG]

### Description

Botany splicer shows some seeds on duplicate which are not splicable or ejectable without eject-everything button. When you eject these seeds one by one, they never leave the splicing machine.

### Steps to reproduce

This is how I was able to reproduce this, not sure if you need every step
1. Get a bunch of different seeds. I used produce satchel to get good 10 of three different species, I am not sure if it matters.
2. Eject them all out, so they are on top of the machine
3. Drag and drop seeds into the splicer. Do multiple drag and drop calls (so drag and drop one, and while others are getting loaded by your previous drag+drop do it again)
4. You are stuck with weird seeds which do not show type, GN or any other stat on the console. You are not able to splice or eject these without some weird visual glitches.

### Expected Behavior

Not having ghost seeds on the console

### Additional Information & Screenshots

image
image

### LINKED ISSUE #7465: PlantMaster UI keeps opening as AI
### Labels

[BUG]

### Description

PlantMaster UI keeps re-opening after being closed

### Steps to reproduce

1. Be AI
2. Click on PlantMaster
3. Close UI
4. Every time someone interacts with the PlantMaster it re-opens on my screen

### Expected Behavior

I expected the UI to only update on the botanist's client and not mine

### Additional Information & Screenshots

This is probably inconsistent as hell, but I'm putting it here because it annoyed me and may be fixable
Also probably not specific to PlantMasters, might be related to #6612 
Workaround is to minimize it instead of close it

### LINKED ISSUE #9339: Plantmaster Forcing Connection Reset
### Labels

[BUG]

### Description

After using the plantmaster for a bit, it stopped working and clicking on it would cause my connection to reset.

### Steps to reproduce

Possibly splicing multiple times will cause this, not entirely sure.

### Expected Behavior

To be functional after splicing plants

### Additional Information & Screenshots

Reported by: NotDucc
On server: Goonstation 4 Roleplay: Sylvester
Round log date: 2022-06-24-23-10
Reported on: 2022-06-25 00:21:45

### LINKED ISSUE #9608: PlantMaster Mk3 Splice Bug?
### Labels

[MAJOR]

### Description

Splicing will just not work fo the Seed and it makes the Console unuseable afterwards.

### Steps to reproduce

1:Put a seed in the PlantMaster.
2:Splice the seed.
3:Seed shouldn't Splice and after that close the console.

### Expected Behavior

This should at least in my case not make the Console be useable again, happens after a few tries too or instantly.

### Additional Information & Screenshots

Reported by: Fredric_80100
On server: Goonstation 4 Roleplay: Sylvester
Round log date: 2022-07-12-21-08
Reported on: 2022-07-12 21:41:20

### LINKED ISSUE #10245: Opening Seed Extractor Freezes Client
### Labels

[BUG][MAJOR]

### Description

I inserted some seeds into the seed extractor machine and now the extractor with the specific seed combination freezes my client whenever I open it.

### Steps to reproduce

Insert seeds, quantity doesn't seem to matter much. Open the seed extractor interface.

### Expected Behavior

Opens the UI without freezing your client for ~20 seconds.

### Additional Information & Screenshots

My network connection to the server stops whenever I open the interface for the remainder of the round. It only occurs with the one I was inserting seeds into. Memory usage also drops from 400M to 190M. Could be related to the general glitchiness of the seed extractor UI? I've seen seeds that don't actually exist appear in the UI but you're able to interact with them.

Reported by: Juandeaglio
On server: Goonstation 1 Classic: Heisenbee
Round log date: 2022-08-13-19-30
Reported on: 2022-08-13 20:29:39

### PR DESCRIPTION
**Pull request by amylizzle**
## About the PR 
Behold, the latest iteration of the Plantmaster series: the Mark 4! Now with multi-reagent infusing and snazzy new interface!

image
image
image
image
image

## Why's this needed? 
TGUI good. Also (hopefully) fixes a bunch of bugs:
Fixes #10245 
Fixes #9608 - pretty sure this one was a consequence of some jank which I fixed
Fixes #9339
Fixes #7465 
Fixes #6429 - pretty sure I fixed this one, I replicated it once, fixed it, and couldn't do it again
Fixes #4974
Fixes #2273

## Changelog

```changelog
(u)Amylizzle
(*)TGUI'd the Plantmaster. Now with multi-agent infusion!
```

### CODE REVIEW

File: tgui/packages/tgui/interfaces/Plantmaster.js [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+              act('change_tab', { 'tab': 'extractables' });
+            }}>
+            Seed Extraction {`(${category_lengths[0]})`}
+          </Tabs.Tab>
+          <Tabs.Tab
+            selected={category === 'seedlist'}
+            onClick={() => {
+              act('change_tab', { 'tab': 'seedlist' });
+            }}>
+            Seeds {`(${category_lengths[1]})`}
```
Comment (Luxizzle) [inline tgui/packages/tgui/interfaces/Plantmaster.js:61]:
```suggestion
            Seeds ({category_lengths[1]})
```

File: tgui/packages/tgui/interfaces/Plantmaster.js [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+      </Window.Content>
+    </Window>
+  );
+};
+
+const compare = function (a, b, sortBy, sortAsc) {
+  if (sortBy === "name" || sortBy === "species")
+  {
+    if (sortAsc) {
+      return (''+a[sortBy]).localeCompare(b[sortBy]);
```
Comment (Luxizzle) [inline tgui/packages/tgui/interfaces/Plantmaster.js:86]: Just a little cleaner instead of using the JS quirks with types. Especially if someone ever wants to move this to typescript at some point

```suggestion
      return (a[sortBy] ?? '').localeCompare(b[sortBy] ?? '');
```

File: tgui/packages/tgui/interfaces/Plantmaster.js [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+        {category === 'extractables' && <PlantExtractables seedoutput={seedoutput} produce={extractables} sortBy={sortBy} sortAsc={sortAsc} />}
+        {category === 'seedlist' && <PlantSeeds seeds={seeds} seedoutput={seedoutput} splicing={show_splicing} splice_chance={splice_chance} splice_seeds={splice_seeds} sortBy={sortBy} sortAsc={sortAsc} />}
+      </Window.Content>
+    </Window>
+  );
+};
+
+const compare = function (a, b, sortBy, sortAsc) {
+  if (sortBy === "name" || sortBy === "species")
+  {
```
Comment (Luxizzle) [inline tgui/packages/tgui/interfaces/Plantmaster.js:84]:
```suggestion
  if (sortBy === "name" || sortBy === "species") {
```

File: tgui/packages/tgui/interfaces/Plantmaster.js [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+    else
+    {
+      return (''+b[sortBy]).localeCompare(a[sortBy]);
+    }
+  }
+  if (sortAsc) {
+    return parseFloat(a[sortBy]) - parseFloat(b[sortBy]);
+  }
+  else
+  {
```
Comment (Luxizzle) [inline tgui/packages/tgui/interfaces/Plantmaster.js:97]:
```suggestion
  else {
```

File: tgui/packages/tgui/interfaces/Plantmaster.js [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+            backgroundColor={inserted_container !== null ? "green" : "blue"}
+            selected={inserted_container !== null}
+            icon="eject"
+            onClick={() => inserted_container !== null ? act('ejectbeaker') : act('insertbeaker')}
+            bold>
+            {inserted_container !== null ? inserted : "Insert Beaker"}
+          </Tabs.Tab>
+        </Tabs>
+
+
```
Comment (Luxizzle) [inline tgui/packages/tgui/interfaces/Plantmaster.js:73]:
```suggestion
```

File: tgui/packages/tgui/interfaces/Plantmaster.js [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+        </Flex.Item>
+      </Flex>
+    </SectionEx>
+  );
+};
+const TitleRow = (props, context) => {
+  const { act } = useBackend(context);
+  const { show_damage, sortBy, sortAsc } = props;
+  const headings = ["name", "species", "damage", "genome", "generation", "maturity rate", "production rate", "lifespan", "yield", "potency", "endurance", "controls"];
+  const sortname = ["name", "species", "damage", "genome", "generation", "growtime", "harvesttime", "lifespan", "cropsize", "potency", "endurance", ""];
```
Comment (Luxizzle) [inline tgui/packages/tgui/interfaces/Plantmaster.js:151]: These can be constants at the top of the file, don't have to create a new array every time this is rendered

File: tgui/packages/tgui/interfaces/Plantmaster.js [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+            currentValue={extractable.name[0]}
+            onCommit={(e, new_name) => act('label', {
+              label_ref: extractable.ref[0],
+              label_new: new_name,
+            })} />
+        </TableCell>
+        <TableCell width="100px" textAlign="center" bold={extractable.species[1]} backgroundColor={extractable.species[1] ? "#333333" : ""}>
+          {extractable.species[0]}
+        </TableCell>
+        { show_damage && (
```
Comment (Luxizzle) [inline tgui/packages/tgui/interfaces/Plantmaster.js:190]:
```suggestion
        {show_damage && (
```

File: tgui/packages/tgui/interfaces/Plantmaster.js [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+          </Button.Checkbox>
+        </Flex.Item>
+      )}>
+      <Flex height="100%" direction="column">
+        <Flex.Item height={splicing ? "60%" : "100%"}>
+          <Table>
+            <TitleRow show_damage sortBy={sortBy} sortAsc={sortAsc} />
+            {extractablesOnPage.map((extractable, index) => (
+              <PlantRow extractable={extractable} key={extractable.ref[1]}
+                show_damage infuse splice splice_disable={splice_disable} />
```
Comment (Luxizzle) [inline tgui/packages/tgui/interfaces/Plantmaster.js:320]:
```suggestion
              <PlantRow 
                extractable={extractable} 
                key={extractable.ref[1]}
                show_damage 
                infuse 
                splice 
                splice_disable={splice_disable} 
              />
```

File: tgui/packages/tgui/interfaces/Plantmaster.js [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+            onClick={() => inserted_container !== null ? act('ejectbeaker') : act('insertbeaker')}
+            bold>
+            {inserted_container !== null ? inserted : "Insert Beaker"}
+          </Tabs.Tab>
+        </Tabs>
+
+
+        {category === 'overview' && <PlantOverview cat_lens={category_lengths} container={inserted ? inserted_container : null} seedoutput={seedoutput} />}
+        {category === 'extractables' && <PlantExtractables seedoutput={seedoutput} produce={extractables} sortBy={sortBy} sortAsc={sortAsc} />}
+        {category === 'seedlist' && <PlantSeeds seeds={seeds} seedoutput={seedoutput} splicing={show_splicing} splice_chance={splice_chance} splice_seeds={splice_seeds} sortBy={sortBy} sortAsc={sortAsc} />}
```
Comment (Luxizzle) [inline tgui/packages/tgui/interfaces/Plantmaster.js:76]: Minor opinionated problem, but these go way beyond 80 characters on a single line and are just hard to parse quickly

This goes for everything in this file, I like to run the prettier formatter first and fix it up with the eslint auto-fix after to fix up any of tgui's eslint rules.

ESLint just kinda sucks at this stuff.

```suggestion
        {category === 'overview' && (
          <PlantOverview
            cat_lens={category_lengths}
            container={inserted ? inserted_container : null}
            seedoutput={seedoutput}
          />
        )}
        {category === 'extractables' && (
          <PlantExtractables seedoutput={seedoutput} produce={extractables} sortBy={sortBy} sortAsc={sortAsc} />
        )}
        {category === 'seedlist' && (
          <PlantSeeds
            seeds={seeds}
            seedoutput={seedoutput}
            splicing={show_splicing}
            splice_chance={splice_chance}
            splice_seeds={splice_seeds}
            sortBy={sortBy}
            sortAsc={sortAsc}
          />
        )}
```

File: code/obj/submachine/seed.dm [hunk 10]
```diff
@@ -1,632 +1,491 @@
...
+			"category_lengths" = list(length(src.extractables),length(src.seeds)),\
+			"inserted" =  src.inserted ? "[src.inserted.reagents.total_volume]/[src.inserted.reagents.maximum_volume] [src.inserted.name]" : "No reagent vessel",\
+			"inserted_container" = thisContainerData,\
+			"seedoutput" = src.seedoutput,\
+			"splice_chance" = splice_chance,\
+			"show_splicing" = src.splicing1 || src.splicing2,
+			"splice_seeds" = list(splice1_geneout, splice2_geneout),
+			"sortBy" = src.sort,
+			"sortAsc" = src.sortAsc,
+		)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:110]:
```suggestion
		return list(
			"extractables" = exlist,
			"seeds" = seedlist,
			"category" = src.mode,
			"category_lengths" = list(length(src.extractables),length(src.seeds)),
			"inserted" =  src.inserted ? "[src.inserted.reagents.total_volume]/[src.inserted.reagents.maximum_volume] [src.inserted.name]" : "No reagent vessel",
			"inserted_container" = thisContainerData,
			"seedoutput" = src.seedoutput,
			"splice_chance" = splice_chance,
			"show_splicing" = src.splicing1 || src.splicing2,
			"splice_seeds" = list(splice1_geneout, splice2_geneout),
			"sortBy" = src.sort,
			"sortAsc" = src.sortAsc,
		)
```

File: code/obj/submachine/seed.dm [hunk 11]
```diff
@@ -1,632 +1,491 @@
...
-			return
-		if(href_list["page"])
-			var/ops = text2num_safe(href_list["page"])
-			switch(ops)
-				if(2) src.mode = "extraction"
-				if(3) src.mode = "seedlist"
-				else src.mode = "overview"
-			playsound(src.loc, "sound/machines/click.ogg", 50, 1)
-			src.updateUsrDialog()
+	ui_act(action, list/params, datum/tgui/ui)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:118]: Missing a lil code here that would otherwise allow you to interact with this machine, even if not in range for example

```suggestion
	ui_act(action, list/params, datum/tgui/ui)
		. = ..()
		if(.)
			return
```

File: code/obj/submachine/seed.dm [hunk 12]
```diff
@@ -1,632 +1,491 @@
...
+		switch(action)
+			if("change_tab")
+				src.mode = params["tab"]
+				playsound(src.loc, "sound/machines/click.ogg", 50, 1)
+				tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:123]: We can actually return `TRUE` here and tgui will handle the update itself! No need to do this ourselves

```suggestion
				return TRUE
```

File: code/obj/submachine/seed.dm [hunk 13]
```diff
@@ -1,632 +1,491 @@
...
+					boutput(usr, "<span class='alert'>No receptacle found to eject.</span>")
 				else
-					I.set_loc(src.loc) // causes Exited proc to be called
-					usr.put_in_hand_or_eject(I) // try to eject it into the users hand, if we can
+					if (I.cant_drop) // cyborg/item arms
+						src.inserted = null
+					else
+						I.set_loc(src.loc) // causes Exited proc to be called
+						usr.put_in_hand_or_eject(I) // try to eject it into the users hand, if we can
+				tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:135]:
```suggestion
				return TRUE
```

File: code/obj/submachine/seed.dm [hunk 14]
```diff
@@ -1,632 +1,491 @@
...
+						boutput(ui.user, "This machine does not accept containers from robots!")
+						return
+					if(src.inserted)
+						boutput(ui.user, "<span class='alert'>A container is already loaded into the machine.</span>")
+						return
+					src.inserted =  inserting
+					ui.user.drop_item()
+					inserting.set_loc(src)
+					boutput(ui.user, "<span class='notice'>You add [inserted] to the machine!</span>")
+					tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:152]:
```suggestion
					return TRUE
```

File: code/obj/submachine/seed.dm [hunk 15]
```diff
@@ -1,632 +1,491 @@
...
+
+			if("ejectseeds")
+				for (var/obj/item/seed/S in src.seeds)
+					src.seeds.Remove(S)
+					S.set_loc(src.loc)
+					usr.put_in_hand_or_eject(S) // try to eject it into the users hand, if we can
+					tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:159]:
```suggestion
				return TRUE
```

File: code/obj/submachine/seed.dm [hunk 16]
```diff
@@ -1,632 +1,491 @@
...
 
-			src.updateUsrDialog()
+			if("ejectextractables")
+				for (var/obj/item/I in src.extractables)
+					src.extractables.Remove(I)
+					I.set_loc(src.loc)
+					usr.put_in_hand_or_eject(I) // try to eject it into the users hand, if we can
+					tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:166]:
```suggestion
				return TRUE
```

File: code/obj/submachine/seed.dm [hunk 17]
```diff
@@ -1,632 +1,491 @@
...
-			else src.extractables.Remove(I)
-			if(I == src.splicing1)
-				src.splicing1 = null
-			if(I == src.splicing2)
-				src.splicing2 = null
-			I.set_loc(src.loc)
-			usr.put_in_hand_or_eject(I) // try to eject it into the users hand, if we can
-			src.updateUsrDialog()
+				ui.user.put_in_hand_or_eject(I) // try to eject it into the users hand, if we can
+				tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:180]:
```suggestion
				return TRUE
```

File: code/obj/submachine/seed.dm [hunk 18]
```diff
@@ -1,632 +1,491 @@
...
-					I.name = newName
-			src.updateUsrDialog()
 
-		else if(href_list["filter"])
-			src.seedfilter = copytext(strip_html(input(usr,"Search for seeds by name? (Enter nothing to clear filter)","[src.name]",null)), 1, 257)
-			src.updateUsrDialog()
+			if("sort")
+				src.sort = params["sortBy"]
+				src.sortAsc = text2num(params["asc"])
+				tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:186]:
```suggestion
				return TRUE
```

File: code/obj/submachine/seed.dm [hunk 19]
```diff
@@ -1,632 +1,491 @@
...
-				else if(stored.no_extract)
-					give = 0
-
-				if (!give)
-					boutput(usr, "<span class='alert'>No viable seeds found in [I].</span>")
+					boutput(ui.user, "<span class='alert'>Item cannot be scanned.</span>")
+
+			if("outputmode")
+				src.seedoutput = !src.seedoutput
+				tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:210]:
```suggestion
				return TRUE
```

File: code/obj/submachine/seed.dm [hunk 20]
```diff
@@ -1,632 +1,491 @@
...
+
+			if("label")
+				var/obj/item/I = locate(params["label_ref"]) in src
+				var/newname = sanitize(strip_html(params["label_new"]))
+				if(istype(I) && I.name != newname)
+					phrase_log.log_phrase("seed", newname, TRUE)
+					I.name = newname
+				tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:218]:
```suggestion
				return TRUE
```

File: code/obj/submachine/seed.dm [hunk 21]
```diff
@@ -1,632 +1,491 @@
...
+
+							S.name = "[seedname] seed"
+
+							S.generation = P.generation
+							if (!src.seedoutput) src.seeds.Add(S)
+							else S.set_loc(src.loc)
+							give -= 1
+					src.extractables.Remove(I)
+					qdel(I)
+					tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:275]:
```suggestion
					return TRUE
```

File: code/obj/submachine/seed.dm [hunk 22]
```diff
@@ -1,632 +1,491 @@
...
+										break
+									if (3)
+										playsound(src, "sound/machines/buzz-sigh.ogg", 50, 1)
+										boutput(usr, "<span class='alert'>ERROR: Unknown error. Please try again.</span>")
+										break
+									else
+										playsound(src, "sound/effects/zzzt.ogg", 50, 1)
+										boutput(usr, "<span class='notice'>Infusion of [R.name] successful.</span>")
+								src.inserted.reagents.remove_reagent(R.id,10)
+					tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:347]:
```suggestion
					return TRUE
```

File: code/obj/submachine/seed.dm [hunk 23]
```diff
@@ -1,632 +1,491 @@
...
+
+				// Now get rid of the old seeds and go back to square one
+				src.seeds.Remove(seed1)
+				src.seeds.Remove(seed2)
+				src.splicing1 = null
+				src.splicing2 = null
+				qdel(seed1)
+				qdel(seed2)
+				src.mode = "seedlist"
+				tgui_process.update_uis(src)
```
Comment (Luxizzle) [inline code/obj/submachine/seed.dm:486]:
```suggestion
				return TRUE
```

**Review by Luxizzle — CHANGES_REQUESTED** [review]
First pass, mostly on the tgui side because smol brain, plant code is something else huh

and here's a frog

3lpzqbg1h5i91(1)

**Comment by trazodont** [comment]
I've been waiting for this one 👀👀👀

File: code/obj/submachine/seed.dm [hunk 24]
```diff
@@ -637,7 +496,8 @@
 			user.drop_item()
 			W.set_loc(src)
 			boutput(user, "<span class='notice'>You add [W] to the machine!</span>")
-			src.updateUsrDialog()
+			tgui_process.update_uis(src)
```
[the remark this answers is not in the document]
Comment (amylizzle) [reply code/obj/submachine/seed.dm:499]: Not this one, it's in `attackby` not `uiact`

File: tgui/packages/tgui/interfaces/Plantmaster.js (hunk 9 above)
Comment (amylizzle) [reply tgui/packages/tgui/interfaces/Plantmaster.js:78]: Honestly that looks so much less readable to me. I'mma leave it

**Comment by TobleroneSwordfish** [comment]
🦀 Plantmaster jank is finally dead! 🦀 
Just some UI review for now:
- the controls buttons on the end could do with being bigger
- the eject all button doesn't eject all
- you probably shouldn't be able to sort by the controls column (not sure if this is a tgui thing)
- the first click on all column headings sets them to sort ascending, should probably be descending
- you still can't swap reagents containers, though I'll just do that in a separate PR if you don't include it here

**Comment by amylizzle** [comment]
!merge_upstream

File: code/obj/submachine/seed.dm [hunk 25]
```diff
@@ -1,632 +1,494 @@
...
-					</thead>
-					<tbody>
-				"}
-
-			if(src.extractables.len)
-				for (var/obj/item/I in src.extractables)
-					var/geneout = ""
-					if (istype(I, /obj/item/seed))
-						var/obj/item/seed/S = I
+	ui_data(mob/user)
```
Comment (ZeWaka) [inline code/obj/submachine/seed.dm:25]: can this not all be static data and then just manually updated whenever state changes?
really don't think we need to be running all of this every second

**Comment by amylizzle** [comment]
oops, sorry about the review requests folks - I beefed a merge and it accidentally added literally everyone as code owners

**Comment by frawhst** [comment]
happens to the best of us

File: code/obj/submachine/seed.dm [hunk 26]
```diff
@@ -1,632 +1,494 @@
...
-
-			if(src.extractables.len)
-				for (var/obj/item/I in src.extractables)
-					var/geneout = ""
-					if (istype(I, /obj/item/seed))
-						var/obj/item/seed/S = I
+	ui_static_data(mob/user)
+		var/exlist = list()
+		var/seedlist = list()
+		var/geneout = null//tmp var for storing analysis results
```
Comment (ZeWaka) [inline code/obj/submachine/seed.dm:28]: zewaka note: /tmp/ might do something i forget if i did that

**Comment by amylizzle** [comment]
!merge_upstream
